src/include/uapi/ipc/control.h

```diff
@@ -165,6 +165,71 @@ struct sof_ipc_ctrl_data {
 	};
 } __attribute__((packed));
 
+/**
+ * Error information reported back for APP_CONFIG_PARAM
+ * (aka VENDOR_CONFIG_PARAM) inside the struct sof_ipc_large_data.
+ */
+struct sof_ipc_large_data_status {
+	uint32_t failed_param_index;	/**< index of the failed parameter */
+	uint32_t app_error_code;	/**< app error code */
+} __attribute__((packed));
+
+/**
+ * Application parameter transferred inside struct sof_ipc_large_data.
+ *
+ * <i>priv_param_id</i> bits are assigned as follows:
+ * - [31-8] : parameter instance,
+ * - [7-0]  : parameter type
+ *
+ * In case of GET operation, specifies maximum parameter size accepted
+ * by the host application. When multiple parameters are tunneled inside
+ * a single request/response, the total size of parameters must fit the
+ * single IPC message payload.
+ */
+struct sof_ipc_large_data_app_param {
+	uint32_t param_id;	/**< parameter id */
+	uint32_t size;		/**< size in bytes */
+	uint32_t data[0];	/**< data */
+} __attribute__((packed));
+
+/**
+ * Large control data.
+ *
+ * <i>attribs</i> bits are assigned as follows:
+ * - [31-30] : reserved
+ * - [29]    : init_block - indicates the first fragment,
+ * - [28]    : final_block - indicates the last fragment,
+ * - [27-20] : param_id - ID of the parameter, 0xFF - APP_CONFIG_PARAM,
```

Comment:
I don't get if there is a link with the priv_param definition and what the bits [27-20] are ? Unclear what this refers to?

Reply:
When the driver wants to get/set a large parameter internally, it sets the `sof_ipc_large_data.attribs.param_id` to one of the unique ids [0,0xFE] specified by the fw uapi.

When the driver receives parameter get/set request from the application, it sets the `sof_ipc_large_data.attribs.param_id` to `APP_CONFIG_PARAM` and serializes the get/set request data to the `sof_ipc_large_data.data` buffer as a series (one or more) of `sof_ipc_large_data_app_param` structures where `sof_ipc_large_data_app_param.param_id` is set to the param id received from the application.

Reply:
It looks like param_id is per component and only has meaningful value for that component ? This looks similar to sof_ipc_ctrl_data which could be extended via reserved fields for total/current block and offset and where param_id is index.

Reply:
@mmaka1 I don't get the concept at all. The driver will not initiate any transaction, it's all done at the request of user-space, or user-space handling controls exposed by the driver based on a topology file.